### DISCUSSION TITLE
How to output MSL shader source code

### DISCUSSION THREAD

**Discussion by XAMPPRocky** [KhronosGroup/MoltenVK#1282] [Q&A] [answered]
Hello I'm working on a tool that uses MoltenVK, however we're running into an issue where the SPIRV we provide is valid, however the MSL that is being generated is not. I'm currently trying to investigate and am having trouble replicating the compiled source code using `spirv-cross` or `MoltenVKShaderConverter`. Is there a way to get the MoltenVK library to output the compiled MSL code it's generating and not just warnings?

**Comment by eddyb** [comment, marked as answer]
There appears to be logging logic in `MVKShaderModule::convert`: https://github.com/KhronosGroup/MoltenVK/blob/82ab40e52f3debfeeac6f27c1be2a270978811f4/MoltenVK/MoltenVK/GPUObjects/MVKShaderModule.mm#L276 https://github.com/KhronosGroup/MoltenVK/blob/82ab40e52f3debfeeac6f27c1be2a270978811f4/MoltenVK/MoltenVK/GPUObjects/MVKShaderModule.mm#L300-L301

**EDIT**: and based on this, what you need to do is set the `MKV_DEBUG=1` environment variable, I believe: https://github.com/KhronosGroup/MoltenVK/blob/84c6095fbf0f562cec08b83b8fff7c359c51aa18/MoltenVK/MoltenVK/Utility/MVKEnvironment.cpp#L30

**Reply by billhollings** [reply]
This answer to use `MVK_DEBUG` is correct. @eddyb Thanks for responding!

Documentation on the configuration environment variables and settings can be found in the `vk_mvk_moltenvk.h` file.
